JOBRUNNER: Move to more mission-like API

jpanikulam opened the pull request:
- Jobs are called the same way missions are
- BagManager creates the bag folder if it doesn't already exist
- There is a bug with bag.dump that appears to have existed before
- C3 behaves a little better

Addresses #140

jpanikulam committed:
SIM: Default turbulence to False

(Sorry Matt :( )

jpanikulam committed:
SIM: Add thruster dead-band simulation

mattlangford committed:
SIM: Move BagManager to python package

jpanikulam committed:
SIM: Make job-runner CLI more like missions

- CLI now uses mission-style runnables instead of independent nodes
- Queue is added

jpanikulam committed:
SIM: Remove extraneous characters in sub.sdf

jpanikulam committed:
SUB: Remove spurious print

jpanikulam committed:
JOBRUNNER: Make bagging optional

- Add CLI switch
- Make BagManager create the "bag" folder if it doesn't already exist

jpanikulam committed:
JOBRUNNER: Add demo job

jpanikulam committed:
SIM: Move align_marker_test

jpanikulam committed:
SIM: Force rectification of down

jpanikulam committed:
SIM: Make channel markers super heavy

mattlangford committed:
JOBRUNNER: Update align_marker_test

jpanikulam committed:
JOBRUNNER: Make test_simple_motion useful

jpanikulam commented:
[And some documentation](https://github.com/uf-mil/Sub8/wiki/Gazebo-Job-Runner)

mattlangford commented on simulation/sub8_gazebo/diagnostics/job_runner.py:
```diff
@@ -1,231 +1,118 @@
+            try:
+                yield current_job.setup()
+            except Exception, error:
+                print "On test #{}, a {} raised an error on test setup".format(
```
Should we log any errors like this or just print them out?

jpanikulam replied:
good point

jpanikulam replied:
We could print everything by default and let the user pipe the output into a file only if they desire?

mattlangford replied:
Sure that works for me.
That would remove the lines below that dump the data to a file as well. Just make sure they are printed instead.

mattlangford commented on simulation/sub8_gazebo/launch/duck.launch:
```diff
@@ -88,5 +88,10 @@
+
   <include file="$(find sub8_launch)/launch/subsystems/cameras/launch/stereo.launch" if="$(arg cameras)"/>
+
+  <group ns="/down/left">
```
Only start if cameras have been set to true?

jpanikulam replied:
roger roger

mattlangford commented on simulation/sub8_gazebo/sub8_gazebo_tools/bag_manager.py:
```diff
@@ -0,0 +1,148 @@
+
+        gen = self.read_from_cache()
+
+        last_timestamps = dict(self.cache_dict)
```
Did adding this not fix it?

jpanikulam replied:
msgs still undefined

mattlangford commented on simulation/sub8_gazebo/diagnostics/job_runner.py:
```diff
@@ -250,55 +136,61 @@ def run_job_loop(self):
+    def queue_job(self, name, runs):
+        available_tests = [test_name for test_name in dir(gazebo_tests) if not test_name.startswith('_')]
+        assert name in available_tests, "Unknown test, {}".format(name)
+        assert isinstance(runs, int), "Cannot do non-integer runs, wtf are you trying to do?"
```
I JUST WANT TO RUN HALF THE MISSION

mattlangford commented on simulation/sub8_gazebo/sub8_gazebo_tools/bag_manager.py:
```diff
@@ -0,0 +1,148 @@
+            msg_name = msg['message_name']
+
+            # Import the message
+            exec("from {0}.msg import {1}".format(msg_type, msg_name))
```
Did you want to change how the imports are done?

jpanikulam replied:
Nah, if you want to do it in the future you can use **import**, which is done in the **init**.py of gazebo_tests

mattlangford commented on simulation/sub8_gazebo/sub8_gazebo_tools/bag_manager.py:
```diff
@@ -0,0 +1,148 @@
+                        continue
+
+                    # Lang -- this doesn't work
+                    if msg_time == last_timestamps[key]:
```
What kind of error is it?

jpanikulam replied:
A few lines later, it fails at msgs.append

I'm not sure if my guessery for what to do with last_timestamps is correct, either; is it?

mattlangford replied:
Yes I think that is right. It just needs to be a dictionary with the same keys as the caching dict so it can add timestamps to that.
It is supposed to dump to a bag here instead of save the messages (I copied this from the start_caching function but obv forgot to change some things).
Try this line instead of 119:
`bag.write(key, (yield msg), t=(yield msg_time))`

jpanikulam replied:
Now getting a lot of "There's a problem recording /odom" printouts. /odom is definitely publishing, hopefully this isn't a weird tx problem

mattlangford replied:
That's happening on line 55 probably unless those printouts happen after there is a crash?
It should only print if the last message is none... hmmmm

mattlangford commented on simulation/sub8_gazebo/diagnostics/gazebo_tests/align_marker_test.py:
```diff
@@ -0,0 +1,85 @@
+from diagnostics.gazebo_tests import common
+
+
+class Job(common.Job):
```
This looks a hell of a lot cleaner then what I originally did

mattlangford commented on simulation/sub8_gazebo/diagnostics/job_runner.py:
```diff
@@ -1,231 +1,118 @@
+        if self.record_log:
+            self.bagger.start_caching()
+
+        while current_loop < loop_count:
```
Should initial setup be called up here somewhere?

jpanikulam replied:
added

jpanikulam committed:
JOBRUNNER: Add verbosity option

jpanikulam committed:
SIM: Address PR#143 comments

mattlangford commented on simulation/sub8_gazebo/launch/duck.launch:
```diff
@@ -2,7 +2,7 @@
   <param name="simulate" value="true" />
```
Hey I noticed I made a goof - this needs to be:
`<param name="/use_sim_time" value="true" />`
I think something used to set that but it has gotten removed.

jpanikulam replied:
Don't worry - I think I wrote that line. /simulate is sub-specific and tells the thruster driver to behave differently. Gazebo automatically sets use_sim_time

mattlangford replied:
Hmm I was having issues with TF without it that were solved by adding the line.
Empty_world.launch would set it but I removed the call to that file. I think we need both.

mattlangford replied:
Can the thruster mapper look for use_sm_time? Was there a reason for both?

jpanikulam replied:
I am having those tf issues too, now. I think it's due to starting out of order.

One reason is that Forrest suggested using sim time even in real operation to force time synchronization

mattlangford replied:
That's an interesting thought, I can't think of any reason not to as long as everything is using it

jpanikulam commented:
Addressed comments, couldn't figure out the /odom thing. In response, I added a --log switch, so you can turn off logging. Maybe you'll be able to solve it.

I decided against forcing the user to pipe output, because I think there are some things we want a log file for, and some things we want miscellaneous output for.
